### PR TITLE
Remove meta device utilities in favor of torchdistx

### LINKED ISSUE #13660: Meta tensors have side effects
### To Reproduce

Remove the standalone decorator as in this commit: https://github.com/Lightning-AI/lightning/pull/13651/commits/25935fefc0b8ad1f9e22e7520cf04768fa29fc03

Run the Meta tests together with the rest
```bash
lightning/tests/tests_pytorch$ pytest utilities/test_meta.py utilities/test_model_summary.py utilities/test_parameter_tying.py
```

There are side effects that cause the following error. The Meta test must be leaving global state modified.

```bash
E       NotImplementedError: Could not run 'aten::_cat' with arguments from the 'Meta' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::_cat' is only available for these backends: [CPU, CUDA, QuantizedCPU, BackendSelect, Python, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradLazy, AutogradXPU, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, AutocastCPU, Autocast, Batched, VmapMode, Functionalize].
```
### Expected behavior

This should not happen. We want to remove the `standalone` mark because it's very expensive in our CI

### Environment

PyTorch 1.10+

cc @borda @akihironitta

### LINKED ISSUE #13796: `init_meta` function return type mismatch
## Incomplete Code

This is related with #13763 PR
and `pytorch_lightning/utilities/meta.py`

### Problem

```python
def init_meta(module_fn: Callable[..., Module], *args, **kwargs) -> Module:
       ...
       return module

class _MaterializerModule(subclass, metaclass=_IsinstanceMetaclass):
       ...
       def __new__(cls, *args, **kwargs):
              subclass = cls.__bases__[0]
              cls.add_subclasses(subclass)
              with cls.instantiation_context()
                      obj = init_meta(subclass, *args, **kwargs)

              obj.materialize = partial(cls.materialize, materialize_fn=obj.materialize)
              return obj
```

1) Return Type
`init_meta` function returns `torch.nn.Module` type. But `__new__` method assumes that `obj`  is `_MaterizerModule` or at least its subclass.

2) Parameter Type Mismatch
 `init_meta` function takes `Callable` as first parameter. But in `__new__` method, `subclass` variable is a class.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @otaj

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See https://github.com/pytorch/torchdistx

Fixes #13796
Fixes #13660
Closes #13763

### Does your PR introduce any breaking changes? If yes, please list them.

Removes the experimental code for meta tensor materialization.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @rohitgr7 @tchaton @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -649,6 +647,7 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         """
         try:
             if self.strategy.launcher is not None:
+                self._check_torchdistx_support()
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:650]: I don't like having to put this here. Does anybody know a better option?

File: requirements/pytorch/test.txt [hunk 2]
```diff
@@ -16,3 +16,5 @@ psutil  # for `DeviceStatsMonitor`
 pandas  # needed in benchmarks
 fastapi
 uvicorn
+--extra-index-url https://download.pytorch.org/whl/cpu
```
Comment (justusschock) [inline requirements/pytorch/test.txt:19]: should we really hardcode the cpu version here?

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (justusschock) [reply src/pytorch_lightning/trainer/trainer.py:650]: what about a property on the launcher? Something like `meta_supported` or `distx_supported` which we can check?

Not an ideal option either, but couldn't think of anything better...

File: requirements/pytorch/test.txt (hunk 2 above)
Comment (carmocca) [reply requirements/pytorch/test.txt:19]: Probably not. I don't know how to set this based on the current setup.

Any ideas @Borda or @akihironitta?

Comment (Borda) [reply requirements/pytorch/test.txt:19]: Not sure what you want to achieve... we shall not limit the dev environment.
We are referring to the CPU version in our CPU-based CI because it just wasting time for download and size or occupied cache to use the general torch version...

File: src/pytorch_lightning/strategies/launchers/multiprocessing.py [hunk 3]
```diff
@@ -164,6 +165,16 @@ def _collect_rank_zero_results(self, trainer: "pl.Trainer", results: Any) -> Opt
 
         return _WorkerOutput(best_model_path, weights_path, trainer.state, results, extra)
 
+    def _check_torchdistx_support(self) -> None:
+        if self._start_method == "spawn":
+            from pytorch_lightning.utilities.meta import _is_deferred
+
+            if _is_deferred(self._strategy.lightning_module):
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/launchers/multiprocessing.py:172]: I think you need to handle `is_deferred(None)`? For Lite the `lightning_module` reference would be None

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1451,20 +1449,14 @@ def _call_setup_hook(self) -> None:
...
+            if _module_available("torchdistx.deferred_init"):
+                from torchdistx.deferred_init import materialize_module
 
-        if isinstance(self.strategy, DDPSpawnStrategy):
-            raise MisconfigurationException("LightningModule on meta device isn't supported with spawn.")
+                materialize_module(self.lightning_module)
 
-        materialize_module(self.lightning_module)
-        # the trainer reference is lost during materialization
-        self.lightning_module.trainer = proxy(self)
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:1467]: I assume this is not the case anymore?

File: src/pytorch_lightning/utilities/meta.py [hunk 5]
```diff
@@ -11,149 +11,46 @@
...
 
-else:
 
-    def init_meta(*_, **__):
-        if not _TORCH_GREATER_EQUAL_1_10:
-            return MisconfigurationException("`init_meta` is supported from PyTorch 1.10.0")
+def get_all_subclasses(cls: Type) -> Set[Type]:
+    rank_zero_deprecation(
+        "`pytorch_lightning.utilities.meta.get_all_subclasses` is deprecated in v1.8 and will be removed in v1.9."
+        " Please copy its implementation if you have an use for it."
```
Comment (awaelchli) [inline src/pytorch_lightning/utilities/meta.py:44]:
```suggestion
        " Please copy its implementation if you have a use for it."
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:1467]: Correct. It's working now

File: .actions/setup_tools.py [hunk 6]
```diff
@@ -41,33 +43,56 @@ def _load_py_module(name: str, location: str) -> ModuleType:
     return py
 
 
-def load_requirements(
-    path_dir: str, file_name: str = "base.txt", comment_char: str = "#", unfreeze: bool = True
-) -> List[str]:
+class _RequirementWithComment(pkg_resources.Requirement):
```
Comment (carmocca) [inline .actions/setup_tools.py:46]: I need to pull these changes into a separate PR. They are necessary here because the load_requirements code breaks with `--extra-index-url https://download.pytorch.org/whl/cpu`

**Comment by carmocca** [comment]
Converting to draft while I address the dependency issues

File: requirements/pytorch/test.txt (hunk 2 above)
Comment (carmocca) [reply requirements/pytorch/test.txt:19]: The index url is different whether you are on a CPU or CUDA machine.

So writing `/cpu` in the requirements is not quite correct. How can we choose dynamically?

**Comment by carmocca** [comment]
Since this is all very experimental, I decided not to add the dependency. It's very tricky choosing the ABI version dynamically. And it's only released with PyTorch 1.12 compatibility.

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
